src/framework/mpe/mpetypes.h

```diff
@@ -0,0 +1,516 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef MU_MPE_MPETYPES_H
+#define MU_MPE_MPETYPES_H
+
+#include <stdint.h>
+#include <math.h>
+#include <algorithm>
+#include <unordered_map>
+#include <map>
+#include <vector>
+
+#include "log.h"
+
+namespace mu::mpe {
+// common
+using msecs_t = uint64_t;
+using percentage_t = float; // from 0.0 (0%) to 1.0 (100%)
+constexpr percentage_t PERCENTAGE_PRECISION_STEP = 0.1; // 10%
+
+// Arrangement
+using timestamp_t = msecs_t;
+using duration_t = msecs_t;
```

Comment:
Maybe it is better to indicate this value in ticks? Then it will not depend on the tempo and, when the tempo changes, it will remain valid, there will be no need to recalculate and update it. For example, this can be important if we want to write the rendered events separately and synchronize them with the notation representation.

Reply:
I was thinking about this, but decided to not introduce any extra terms/conversions here. Because, in fact, I'll need to recalculate and update data anyway, even using an additional indexing. For example, if you change the duration of the note, you have to shift indexes of the rest of events. Also the data should be updated for any dynamic change as well.

So I decided to stick with milliseconds for now, since it allows to avoid unnecessary conversions, provides a more convenient way to handle complex articulations and it's a bit clearer to understand.